client/network/src/request_responses.rs

```diff
@@ -114,8 +115,27 @@ pub struct IncomingRequest {
 	/// [`ProtocolConfig::max_request_size`].
 	pub payload: Vec<u8>,
 
-	/// Channel to send back the response to.
-	pub pending_response: oneshot::Sender<Vec<u8>>,
+	/// Channel to send back the response.
+	///
+	/// There are two ways to indicate that handling the request failed:
+	///
+	/// 1. Drop `pending_response` and thus not changing the reputation of the peer.
+	///
+	/// 2. Sending an `Err(())` via `pending_response`, optionally including reputation changes for
+	/// the given peer.
+	pub pending_response: oneshot::Sender<Response>,
+}
+
+/// Response for an incoming request to be send by a request protocol handler.
+#[derive(Debug)]
+pub struct Response {
+	/// The payload of the response.
+	///
+	/// `Err(())` if none is available e.g. due an error while handling the request.
+	pub result: Result<Vec<u8>, ()>,
+	/// Reputation changes accrued while handling the request. To be applied to the reputation of
+	/// the peer sending the request.
+	pub reputation_changes: Option<Vec<ReputationChange>>,
```

Comment:
```suggestion
	pub reputation_changes: Vec<ReputationChange>,
```

I don't understand why this `Option` is needed, since `None` and `Vec::new()` represent the same thing.

Reply:
Good point. Somehow thought `Vec::new` allocates. Done in https://github.com/paritytech/substrate/pull/7958/commits/3abb27a49da065e44f8d4ce03a62286e23c3d021.